src/HDWallet.cpp

```diff
@@ -319,31 +321,36 @@ HDNode getNode(const HDWallet& wallet, TWCurve curve, const DerivationPath& deri
     auto node = getMasterNode(wallet, curve);
     for (auto& index : derivationPath.indices) {
         switch (privateKeyType) {
-            case HDWallet::PrivateKeyTypeDoubleExtended: // used by Cardano, special handling
-                hdnode_private_ckd_cardano(&node, index.derivationIndex());
-                break;
-           case HDWallet::PrivateKeyTypeDefault32:
-            default:
-                hdnode_private_ckd(&node, index.derivationIndex());
-                break;
+        case HDWallet::PrivateKeyTypeDoubleExtended:
+            hdnode_private_ckd_cardano(&node, index.derivationIndex());
+            break;
+        case HDWallet::PrivateKeyTypeDefault32:
+        default:
+            hdnode_private_ckd(&node, index.derivationIndex());
+            break;
         }
     }
     return node;
 }
 
 HDNode getMasterNode(const HDWallet& wallet, TWCurve curve) {
     const auto privateKeyType = HDWallet::getPrivateKeyType(curve);
-    auto node = HDNode();
+    HDNode node;
     switch (privateKeyType) {
-        case HDWallet::PrivateKeyTypeDoubleExtended: // used by Cardano
-            // special handling for extended, use entropy (not seed)
-            hdnode_from_entropy_cardano_icarus((const uint8_t*)"", 0, wallet.getEntropy().data(), (int)wallet.getEntropy().size(), &node);
-            break;
-
-        case HDWallet::PrivateKeyTypeDefault32:
-        default:
-            hdnode_from_seed(wallet.getSeed().data(), HDWallet::seedSize, curveName(curve), &node);
-            break;
+    case HDWallet::PrivateKeyTypeDoubleExtended: {
+        // Derives the root Cardano HDNode from a passphrase and the entropy encoded in
+        // a BIP-0039 mnemonic using the Icarus derivation (V2) scheme
+        const auto entropy = wallet.getEntropy();
+        uint8_t secret[CARDANO_SECRET_LENGTH];
```

Comment:
Are we sure there's no possibility for buffer overflow here? Looks like we're sending this buffer around

Reply:
This should be fine here, interacting with trezor c code

Reply:
Lgtm but @Milerius, can you please also verify (as it is a critical spot)?

Reply:
Fine, but an `assert()` on the size could make sense

Reply:
`assert` won't compile in release mode